src/tree/updater_colmaker.cc

```diff
@@ -254,13 +246,13 @@ class ColMaker: public TreeUpdater {
     // this function does not support nested functions
     inline void ParallelFindSplit(const SparsePage::Inst &col,
                                   bst_uint fid,
-                                  const DMatrix &fmat,
```

Comment:
Why the change from `const DMatrix &fmat` to `DMatrix *p_fmat` here?

Reply:
The function p_fmat->GetColDensity() lazily generates and stores the column densities on demand, therefore on the first call this modifies data in dmatrix.